### PR TITLE
接入pixiu性能测试以及压力测试

### PR DESCRIPTION
**Pull request by kaori-seasons**
1.支持provider-> pixiu -> consumer性能测试统计指标
2.支持provider直连consumer
3.支持http -> dubbo协议转换下 provider->pixiu->consumer 测试

### CODE REVIEW

File: dubbogo/pkg/PerformanceTester.go [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+const (
+	PayloadType_COMPRESSABLE PayloadType = 0
+)
+
+type PayloadType int32
+
+type StressRequest struct {
+	ResponseType PayloadType `json:"response_type,omitempty"`
+	// Desired payload size in the response from the server.
+	ResponseSize int32 `json:"response_size,omitempty"`
```
Comment (baerwang) [inline dubbogo/pkg/PerformanceTester.go]: 注释写在字段旁边

```go
ResponseSize int32 `json:"response_size,omitempty"` // Desired payload size in the response from the server.
```

File: dubbogo/pkg/PerformanceTester.go [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+	// Optional input payload sent along with the request.
+	Payload *Payload `json:"bytes,3,opt,name=payload,proto3" json:"payload,omitempty"`
+}
+
+type Payload struct {
+
+	// 请求包体的大小
+	Type PayloadType `json:"type,omitempty"`
+	// 包体序列化成字节数组的大小
+	Body []byte `json:"body,omitempty"`
```
Comment (baerwang) [inline dubbogo/pkg/PerformanceTester.go]: 同上，翻译英文

File: dubbogo/stress/client.go [hunk 3]
```diff
@@ -0,0 +1,35 @@
+package main
+
+import (
+	"context"
+	"dubbo-go-pixiu-benchmark/dubbogo/main/pkg"
+	"dubbo.apache.org/dubbo-go/v3/config"
+	"fmt"
+	hessian "github.com/apache/dubbo-go-hessian2"
```
Comment (baerwang) [inline dubbogo/stress/client.go:11]: 按照dubbogo import 规范分开下

File: dubbogo/stress/main.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
+package main
+
+import (
+	"dubbo-go-pixiu-benchmark/dubbogo/pkg"
+	"dubbo-go-pixiu-benchmark/stats"
+	"dubbo.apache.org/dubbo-go/v3/common/logger"
+	"flag"
+	"sync"
+	"time"
```
Comment (baerwang) [inline dubbogo/stress/main.go:18]: 同上

File: dubbogo/stress/main.go [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+		Payload: &pkg.Payload{
+			Type: pkg.PayloadType_COMPRESSABLE,
+			Body: make([]byte, *rqSize),
+		},
+	}
+
+	//FIXME 压测器解析dubbo协议请求连接 执行请求
+	r := req
+	if r == nil {
+
```
Comment (baerwang) [inline dubbogo/stress/main.go:110]: result ？

**Review by baerwang — CHANGES_REQUESTED** [review]

File: dubbogo/stress/main.go (hunk 5 above)
Comment (kaori-seasons) [reply dubbogo/stress/main.go:110]: @baerwang 需要参考 https://github.com/apache/dubbo-go-pixiu/blob/develop/pkg/filter/network/dubboproxy/filter/proxy/proxyfilter.go

目前在本地做适配

File: dubbogo/stress/client.go (hunk 3 above)
Comment (kaori-seasons) [reply dubbogo/stress/client.go:11]: 了解 预计下周改动

Comment (ztelur) [reply dubbogo/stress/client.go:11]: 社区推荐代码的 import 块按照所属进行拆分，有一个工具可以自动将import拆分，地址：https://github.com/dubbogo/tools
命令是 imports-formatter -path . -module github.com/apache/dubbo-go-pixiu -bl false
具体的module名称要相应的进行修改

File: pixiu/dubbo/test/pixiu_test.go [hunk 6]
```diff
@@ -0,0 +1,32 @@
+package test
+
+import (
+	"github.com/stretchr/testify/assert"
```
Comment (AlexStocks) [inline pixiu/dubbo/test/pixiu_test.go]: split it

**Review by ztelur — APPROVED** [review]

File: pixiu/triple/go-server/cmd/server.go [hunk 7]
```diff
@@ -0,0 +1,55 @@
+package main
+
+import (
+	"dubbo-go-pixiu-benchmark/api"
+
```
Comment (cityiron) [inline pixiu/triple/go-server/cmd/server.go]: split

File: pixiu/triple/go-client/cmd/client.go [hunk 8]
```diff
@@ -0,0 +1,70 @@
+package main
+
+import (
+	"context"
+	"sync"
+)
+
+import (
+	"dubbo-go-pixiu-benchmark/api"
```
Comment (cityiron) [inline pixiu/triple/go-client/cmd/client.go]: split

File: pixiu/dubbo/go-server/pkg/wrapper_array_class_provider.go [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+	"context"
+)
+
+import (
+	"dubbo.apache.org/dubbo-go/v3/common/logger"
+
+	hessian "github.com/apache/dubbo-go-hessian2"
+)
+
+type WrapperArrayClassProvider struct {
```
Comment (cityiron) [inline pixiu/dubbo/go-server/pkg/wrapper_array_class_provider.go]: Need some comment for method like InvokeWithJavaIntegerArray

File: pixiu/dubbo/go-server/pkg/user.go [hunk 10]
```diff
@@ -0,0 +1,109 @@
+/*
```
Comment (cityiron) [inline pixiu/dubbo/go-server/pkg/user.go:1]: May be all files need this license, consistent with the project configuration.

File: pixiu/dubbo/go-server/pkg/complex_provider.go [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+}
+
+func (u *ComplexData) JavaClassName() string {
+	return "org.apache.dubbo.ComplexData"
+}
+
+type ComplexProvider struct {
+}
+
+func (u *ComplexProvider) InvokeWithMultiBasicData(ctx context.Context, str string, data []byte, num int32, boolValue bool) (int32, error) {
```
Comment (cityiron) [inline pixiu/dubbo/go-server/pkg/complex_provider.go]: Need some comment?

File: pixiu/dubbo/go-server/cmd/server.go [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+import (
+	"fmt"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+)
+
+import (
+	"dubbo-go-pixiu-benchmark/pixiu/dubbo/go-server/pkg"
```
Comment (cityiron) [inline pixiu/dubbo/go-server/cmd/server.go]: split

File: pixiu/dubbo/go-client/cmd/client.go [hunk 13]
```diff
@@ -0,0 +1,102 @@
+package client
+
+import (
+	"context"
+	"time"
+)
+
+import (
+	"dubbo-go-pixiu-benchmark/pixiu/dubbo/go-client/pkg"
```
Comment (cityiron) [inline pixiu/dubbo/go-client/cmd/client.go]: split

File: README.md [hunk 14]
```diff
@@ -1,2 +1,261 @@
...
+50%  285.18ms
+75%  285.28ms
+90%  285.28ms
+99%  285.28ms
+3 requests in 3.09s, 600.00B read  (3.09s内处理了3个请求，耗费流量600.00B)
+Requests/sec:      0.97     (QPS 0.97 即平均每秒数处理请求数0.97  )
+Transfer/sec:     194.20B   (平均每秒流量194.20B )
+```
+
+结果： 每个线程的平均延迟在285.15ms, 每秒请求数 9个/s，延迟分布在t99的 主要延迟时长是285.28ms
```
Comment (cityiron) [inline README.md]: Result highlight is better,  I think add some independent blocks looks better.

File: go.mod [hunk 15]
```diff
@@ -0,0 +1,132 @@
+module dubbo-go-pixiu-benchmark
+
+require (
+	//github.com/apache/dubbo-go-pixiu v0.5.0-rc1
```
Comment (baerwang) [inline go.mod]: del

File: latency/latency.go [hunk 16]
```diff
@@ -0,0 +1,276 @@
...
+type ContextDialer func(ctx context.Context, network, address string) (net.Conn, error)
+
+type Network struct {
+	Kbps    int           // Kilobits per second; if non-positive, infinite
+	Latency time.Duration // One-way latency (sending); if non-positive, no delay
+	MTU     int           // Bytes per packet; if non-positive, infinite
+}
+
+var (
+	//Local simulates local network.
```
Comment (baerwang) [inline latency/latency.go]: space is required

File: latency/latency.go [hunk 17]
```diff
@@ -0,0 +1,276 @@
...
+	var ping [8]byte
+	if err := binary.Read(c.Conn, binary.BigEndian, &ping); err != nil {
+		return err
+	} else if string(ping[:]) != pingMsg {
+		return fmt.Errorf("malformed handshake message: %v (want %q)", ping, pingMsg)
+	}
+
+	// Both sides are alive and syncing.  Calculate network delay / clock skew.
+	att := 0
+	good := 0
```
Comment (baerwang) [inline latency/latency.go]:
```go
	var (
		att, good             int
		localDone, remoteDone bool
		latency               time.Duration
	)
```

**Comment by kaori-seasons** [comment]
The revision is expected to be completed this weekend

**Comment by kaori-seasons** [comment]
Hello everyone, when I recently read the code of banyandb, I found that his server can be started and stopped based on ```onsi/gomega/gexec```. I think this is very helpful for me to run tests after connecting to Github Action, which is equivalent to running in Github Action There is no need to manually start a server locally.
stress_test.go : https://github.com/apache/skywalking-banyandb/blob/main/test/stress/stress_suite_test.go

File: go.mod [hunk 18]
```diff
@@ -0,0 +1,136 @@
+module dubbo-go-pixiu-benchmark
```
Comment (AlexStocks) [inline go.mod:1]: 这个文件分这么多 require 干嘛？这是什么风格？

File: helpers/grpc_health.go [hunk 19]
```diff
@@ -0,0 +1,40 @@
+package helpers
+
+import (
```
Comment (AlexStocks) [inline helpers/grpc_health.go:3]: format it pls

Comment (kaori-seasons) [reply helpers/grpc_health.go:3]: solved

File: go.mod (hunk 18 above)
Comment (kaori-seasons) [reply go.mod:1]: @AlexStocks 因为怕包太多的情况下 全放到一起可能不太合适
https://github.com/apache/skywalking-banyandb/blob/main/go.mod
目前已经参考这里 将dubbo-go的协议包和用到的工具包分成两块

File: latency/latency.go (hunk 17 above)
Comment (kaori-seasons) [reply latency/latency.go]: this class is deprecated

File: latency/latency.go (hunk 16 above)
Comment (kaori-seasons) [reply latency/latency.go]: the class is deprecated

File: go.mod (hunk 15 above)
Comment (kaori-seasons) [reply go.mod]: the class is deprecated

File: pixiu/dubbo/go-server/pkg/wrapper_array_class_provider.go (hunk 9 above)
Comment (kaori-seasons) [reply pixiu/dubbo/go-server/pkg/wrapper_array_class_provider.go]: the class is deprecated

File: pixiu/dubbo/go-server/pkg/complex_provider.go (hunk 11 above)
Comment (kaori-seasons) [reply pixiu/dubbo/go-server/pkg/complex_provider.go]: the class is deprecated

File: pixiu/dubbo/go-server/pkg/user.go (hunk 10 above)
Comment (kaori-seasons) [reply pixiu/dubbo/go-server/pkg/user.go:1]: This file is migrated from apache/dubbo-go-pixiu, so you need to add a license

File: helpers/grpc_client.go [hunk 20]
```diff
@@ -0,0 +1,57 @@
+package helpers
+
+import (
+	"context"
+	"time"
```
Comment (baerwang) [inline helpers/grpc_client.go:5]: split

File: logger/logger.go [hunk 21]
```diff
@@ -0,0 +1,48 @@
+package logger
+
+import (
+	"context"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog"
```
Comment (baerwang) [inline logger/logger.go:10]: split

File: helpers/grpc_client.go (hunk 20 above)
Comment (kaori-seasons) [reply helpers/grpc_client.go:5]: solved

File: logger/logger.go (hunk 21 above)
Comment (kaori-seasons) [reply logger/logger.go:10]: solved

**Review by PhilYue — APPROVED** [review]

**Review by AlexStocks — APPROVED** [review]

**Pull request merged**